packages/drivers/routerlicious-driver/src/test/r11sSocketTests.spec.ts

```diff
@@ -69,129 +74,224 @@ describe("R11s Socket Tests", () => {
+	}
 
-		await assert.rejects(
-			mockSocket(socket as unknown as Socket, async () =>
-				documentService.connectToDeltaStream(client),
-			),
-			{
-				errorType: RouterliciousErrorTypes.clusterDrainingError,
-				scenarioName: "connect_document_error",
+	// Socket error tests
+	const socketErrorTests = [
+		{
+			description: "Socket error with Token Revoked error",
+			errorToThrow: {
+				code: 403,
+				message: "TokenRevokedError",
+				retryAfterMs: 10,
+				internalErrorCode: "TokenRevoked",
+				errorType: DriverErrorTypes.authorizationError,
+				canRetry: false,
+			},
+			expectedErrorType: DriverErrorTypes.authorizationError,
+			expectedInternalErrorCode: "TokenRevoked",
+		},
+		{
+			description: "Socket error with Cluster Draining error",
+			errorToThrow: {
+				code: 503,
+				message: "ClusterDrainingError",
+				retryAfterMs: 1000,
 				internalErrorCode: R11sServiceClusterDrainingErrorCode,
+				errorType: RouterliciousErrorTypes.clusterDrainingError,
+				canRetry: true,
 			},
-			"Error should have occurred",
-		);
-	});
+			expectedErrorType: RouterliciousErrorTypes.clusterDrainingError,
+			expectedInternalErrorCode: R11sServiceClusterDrainingErrorCode,
+		},
+	];
 
-	it("Socket error with Cluster Draining error", async () => {
-		const errorToThrow = {
-			code: 503,
-			message: "ClusterDrainingError",
-			retryAfterMs: 1000,
-			internalErrorCode: R11sServiceClusterDrainingErrorCode,
-			errorType: RouterliciousErrorTypes.clusterDrainingError,
-			canRetry: true,
-		};
-		const errorEventName = "connect_document_success";
+	socketErrorTests.forEach((test) =>
+		runSocketErrorTest(
+			test.description,
+			test.errorToThrow,
+			test.expectedErrorType,
+			test.expectedInternalErrorCode,
+		),
+	);
+
+	// Client error tests
+	runClientErrorTest(
+		"Client Data Corruption error",
+		new DataCorruptionError("Data corruption error", { driverVersion: "1.0" }),
+		FluidErrorTypes.dataCorruptionError,
+	);
+
+	runClientErrorTest(
+		"Client Data Processing error",
+		DataProcessingError.create("Data processing error", "test", undefined, {
+			driverVersion: "1.0",
+		}),
+		FluidErrorTypes.dataProcessingError,
+	);
+
+	it("should not fire abnormal_disconnect event when disconnection happens without an error", async () => {
+		const socketEventName = "connect_document_success";
 		socket = new ClientSocketMock({
-			connect_document: { eventToEmit: errorEventName },
+			connect_document: { eventToEmit: socketEventName },
 		});
-
 		const connection = await mockSocket(socket as unknown as Socket, async () =>
 			documentService.connectToDeltaStream(client),
 		);
-		let error: IAnyDriverError | undefined;
-		connection.on("disconnect", (reason: IAnyDriverError) => {
-			error = reason;
-		});
 
-		// Send Token Revoked error
-		socket.sendErrorEvent(errorToThrow);
+		// Set up a promise that will resolve if abnormal_disconnect is called
+		const disconnectEventP = new Promise<{ clientId: string; errorType: string }>(
```

Comment:
Nit: invert the resolve/reject behaviors in the Promise? I think the test reads a bit more naturally when this reject() means "we received the event we weren't expecting".